### ISSUE TITLE
Encode XML entities rather than wrapping in CDATA

### ISSUE THREAD

**Issue by nikku** [bpmn-io/moddle-xml#31]
There is a number of good reasons [not to use CDATA](https://github.com/camunda/camunda-modeler/issues/760#issuecomment-376608738) for entity encoding. Because of that we should consider encoding entities rather than wrapping text in CDATA section:

__Input__:

```xml
<doc xmlns="http://foo">
&lt;specialChar&gt;
</doc>
```

__Output__:

```xml
<doc xmlns="http://foo"><![CDATA[
<specialChar>
]]></doc>
```

__Expected__:

* preserve encoding _or_ switch to encoding per default

**Comment by nikku** [comment]
Closed via https://github.com/bpmn-io/moddle-xml/commit/5645b5822644a461eba9f3da481362475f040984.

**Comment by falko** [comment]
@nikku Thank you for the quick fix!

**Comment by xzxiaoshan** [comment]
I think it should be up to developers and users to decide whether to use CDATA, neither forcing the use of CDATA nor forcing coding
@falko @nikku

**Comment by xzxiaoshan** [comment]
The forced removal of CDATA has actually caused problems for more developers. This should be a choice decision, not a default forced decision.
[https://forum.bpmn.io/search?q=CDATA](https://forum.bpmn.io/search?q=CDATA)

**Comment by nikku** [comment]
> I think it should be up to developers and users to decide whether to use CDATA, neither forcing the use of CDATA nor forcing coding

@xzxiaoshan Could you elaborate how this caused you issues? Both encoding and CDATA section do the same thing, so I'd like to understand how this caused you troubles?

**Comment by xzxiaoshan** [comment]
[>](url) > I think it should be up to developers and users to decide whether to use CDATA, neither forcing the use of CDATA nor forcing coding
> 
> [@xzxiaoshan](https://github.com/xzxiaoshan) Could you elaborate how this caused you issues? Both encoding and CDATA section do the same thing, so I'd like to understand how this caused you troubles?

@nikku 

Hi,

Thank you for your reply and for asking for more details. Here are my points regarding the necessity of supporting CDATA, based on my real-world experience:

1. I reported a bug to Activiti 1 （https://github.com/Activiti/Activiti/issues/5060） , which is also the base for Camunda. Although I don't have an environment to verify if Camunda has the same issue, the root cause is relevant to both.

2. My actual use case is not Camunda, but a custom application where I need to extend and define custom nodes. For example, I use <custom:script> elements to store scripts.

3. Due to the Activiti bug, the content inside my script tags gets concatenated into a single line, losing all line breaks, and there is no workaround unless the official team fixes it quickly.

4. Initially, I used moddle-xml version 6.0.0 or earlier, and my design was based on using CDATA, which is more suitable for scripts. After upgrading bpmn-js, CDATA was no longer supported, and that's when I discovered the line break issue.

5. After discussing with my team, we believe that while fixing the Activiti bug would solve my immediate problem, from an XML specification perspective, CDATA is a standard and valid design. moddle-xml should not force users to avoid it.

6. I read your issue description and I agree with your point: moddle-xml should not forcefully add CDATA, but it also shouldn't prevent users from choosing to use CDATA if they want. Would you agree?

7. I can't speak for all developers, but I believe many choose CDATA for good reasons. For me, using CDATA means my script content in XML is not encoded, making it much more readable and easier to maintain.

8. Therefore, I think the old approach (before 6.0.0) of always adding CDATA is not ideal, but the new approach (after 6.0.0) of completely disallowing CDATA is also problematic.

9. The best solution would be to let users decide. For example, a quick fix could be: if the content starts with <![CDATA[, then don't escape the body; otherwise, use the default escapeBody. This would maintain compatibility and flexibility.

10. When I need to use CDATA, I can manually wrap the content, e.g., <![CDATA[var a = 10; var b = 10;]]>.

11. As for the potential issue of having ]]> inside the content, I think it's the developer's responsibility to ensure it doesn't happen. If a developer chooses to use CDATA, they should make sure their content doesn't contain ]]>, right?

In summary, I believe moddle-xml should provide the flexibility for users to choose whether to use CDATA or not, rather than enforcing a single approach. This would accommodate more real-world scenarios and developer needs.

---

Determine whether the content is marked with<! [CDATA [Deciding whether to perform escape or not, what do you think about this handling method? Its compatibility is also OK and does not affect all developers who use it.

---

Thank you for considering my feedback!

**Comment by nikku** [comment]
@xzxiaoshan [CDATA](https://www.geeksforgeeks.org/html/xml-cdata-sections/) and [entity encoding](https://www.baeldung.com/xml-encode-special-characters) are two ways to accomplish the exact same thing: Ensure that _any content_ can be put into an XML document, in a way that can be written and read (again). Any reader should support both formats, and any writer should be able to _choose_ in which format to write.

A reader, specifically will read whatever contents (i.e. CDATA encoded) to the actual string representation. They need to provide additional magic to _preserve_ this "it has been encoded using CDATA" so that `CDATA in -> CDATA out` during serialization. This is overhead that is completely unneeded if we do proper entity encoding.

https://github.com/Activiti/Activiti/issues/5060 is clearly an issue on the Activiti side that has to be fixed. Once fixed then your newlines will be preserved, and you'll be able to go with either CDATA or entity encoding just fine.

> For me, using CDATA means my script content in XML is not encoded, making it much more readable and easier to maintain.

We don't optimize for "a human being able to read XML like a book". If that was our target, then CDATA escaping would be superior. I agree. But as said, it is not our target.

**Comment by xzxiaoshan** [comment]
@nikku 

Thank you for your explanation, I understand your point.
So, is there any way for me to preserve CDATA when using moddle-xml right now?

If a developer really needs to preserve CDATA, we cannot assume the needs of all developers—after all, there have always been people discussing this requirement.
If there is indeed a need to use CDATA, is there any way to achieve this?

**Comment by nikku** [comment]
> If there is indeed a need to use CDATA, is there any way to achieve this?

No, there is no way to accomplish this and we also [don't see a need for it](https://github.com/bpmn-io/moddle-xml/issues/31#issuecomment-2995823000). You can resort to older versions of moddle-xml.

**Comment by xzxiaoshan** [comment]
@nikku 
Okay, thank you.
I have submitted a PR for Activiti. Do you have permission to merge Activiti? Can you handle my PR in Activiti?

**Comment by xzxiaoshan** [comment]
https://github.com/Activiti/Activiti/pull/5035   ->  8.1.x

https://github.com/Activiti/Activiti/pull/5061  ->  develop

**Comment by nikku** [comment]
@xzxiaoshan I cannot help you with activity merges. Will close thread. We discussed conclusively I believe.

**Comment by falko** [comment]
@xzxiaoshan you wrote:
> [...] Activiti, which is also the base for Camunda.

Activiti's modeling tools are NOT the base for Camunda's modelling tool and there are [many known issues with the BPMN XML that Activiti produces](https://github.com/camunda-consulting/camunda-7-code-examples/tree/main/snippets/activiti-migration#analyze-and-fix-problematic-bpmn-files).
